code/modules/spatial_hashing.dm

```diff
@@ -0,0 +1,138 @@
+//this is designed for sounds - but maybe could be adapted for more collision / range checking stuff in the future
+
+
+#define GET_NEARBY(A,range) spatial_z_maps[A.z].get_nearby(A,range)
+
+#define CELL_POSITION(X,Y) clamp(((round(X / cellsize)) + (round(Y / cellsize)) * cellwidth) + 1,0,hashmap.len)
+
+#define ADD_BUCKET(X,Y) do{\
+var/cellposition = CELL_POSITION(X,Y);\
+buckets_holding_atom[cellposition] = 1;\
+} while (false)
+
+var/global/list/datum/spatial_hashmap/spatial_z_maps
+
+/proc/init_spatial_map()
+	spatial_z_maps = list(world.maxz)
+	for (var/zlevel = 1; zlevel <= world.maxz; zlevel++)
+		spatial_z_maps[zlevel] = new/datum/spatial_hashmap(world.maxx,world.maxy,60)
+
+/datum/spatial_hashmap
+	var/list/hashmap
+	var/cols
+	var/rows
+	var/width
+	var/height
+	var/cellsize
+	var/cellwidth
+
+	var/last_update = 0
+
+	var/tmp/list/buckets_holding_atom
+	var/tmp/min_x = 0
+	var/tmp/min_y = 0
+	var/tmp/max_x = 0
+	var/tmp/max_y = 0
+
+
+
+	New(w,h,cs)
+		cols = w / cs
+		rows = h / cs
+
+		hashmap = list()
+		hashmap.len = cols * rows
+
+		for (var/i = 1; i <= cols*rows; i++)
+			hashmap[i] = list()
+
+		width = w
+		height = h
+		cellsize = cs
+
+		cellwidth = width/cellsize
+
+		buckets_holding_atom = list()
+		buckets_holding_atom.len = hashmap.len
+
+	/* unused, could be useful later idk
+
+	proc/clear()
+		for (var/i = 1; i <= cols*rows; i++)
+			hashmap[i].len = 0
+
+	proc/register(var/atom/A) //see comments re : single cell
+		for(var/id in get_atom_id(A))
+			hashmap[id] += A;
+	*/
+
+	proc/update()
+		last_update = world.time
+		for (var/i = 1; i <= cols*rows; i++) //clean
+			hashmap[i].len = 0
+		for (var/client/C) //register
+			if (C.mob)
+				hashmap[CELL_POSITION(C.mob.x,C.mob.y)] += C.mob
+				//a formal spatial map implementation would place an atom into any bucket its bounds occupy (register proc instead of the above line). We don't need that here
+				//register(C.mob)
+				//C.mob.maptext = "[CELL_POSITION(C.mob.x,C.mob.y)]" //lazy debug to see what cell we are being placed in
+
+
+
+	proc/get_atom_id(var/atom/A,var/atomboundsize = 33)
+		//usually in this kinda collision detection code you'd want to map the corners of a square....
+		//but this is for our sounds system, where the shapes of collision actually resemble a diamond
+		//so : sample 8 points around the edges of the diamond shape created by our atom
+
+		//N,W,E,S
+		min_x = A.x - atomboundsize
+		min_y = A.y - atomboundsize
+		max_x = A.x + atomboundsize
+		max_y = A.y + atomboundsize
+		ADD_BUCKET(min_x,A.y)
+		ADD_BUCKET(max_x,A.y)
+		ADD_BUCKET(A.x,min_y)
+		ADD_BUCKET(A.x,max_y)
+
+		//NW,NE,SW,SE
+		min_x = A.x - (atomboundsize * 0.7071)
```

Comment:
what is this constant

Reply:
lazy sqrt to get the 'edge' point of a diamond shape diagonally

Reply:
`(sqrt(2)/2)`

Reply:
If performance is an issue and we're doing rounding on the result, a constant is going to be better than using the sqrt function.

Reply:
@Kyle2143 sqrt(constant) is done at compilation